# Exploring eBay Car Sales Data
The aim of the project is to explore and analyze used car listings from eBay Kleinanzeigen *(classifieds section of the German eBay website).*

### Initial Cleaning 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

**Observations**
 - The dataset contains 20 columns, most of which are strings.
 - Some columns have null values, but none have more than ~20% null values.
 - The num_photos column looks odd, we'll need to investigate this further

**Cleaning column labels** - from camelcase to py's preferred snakecase

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [6]:
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  year_of_registration    gearbox  \
0    Angebot  $5,000  control          bus                  2004    manuell   
1    Angebot  $8,500  control    limousine                  1997  automatik   

   power_ps   model   odometer  month_of_registration fuel_type    brand  \
0       158  andere  150,000km                      3       lpg  peugeot   
1       286     7er  150,000km                      6    benzin      bmw   

  not_repaired_damage         date_created  nr_of_pictures  postal_code  \
0                nein  2016-03-26 00:00:00               0        79588   
1                nein  2016-04-04 00:00:00               0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox      power_ps  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

* Dropping column nr_of_pictures, which has all null values 
* Cleaning and converting columns price and odometer from string to float

In [8]:
autos.drop('nr_of_pictures', axis=1)
autos["price"] = (autos["price"]
                  .str.replace("$","")
                  .str.replace(',','')
                  .astype(float))

In [9]:
autos["odometer"] = (autos["odometer"]
                     .str.replace("km","")
                     .str.replace(',','')
                     .astype(float))

In [10]:
autos["odometer"].head()

0    150000.0
1    150000.0
2     70000.0
3     70000.0
4    150000.0
Name: odometer, dtype: float64

In [11]:
autos.rename({"price":"price_dollar","odometer":"odometer_km"}, 
             axis=1, 
             inplace=True)


###  Exploring 'prices_dollar' and 'odometer_km'

In [12]:
autos["price_dollar"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64

In [13]:
autos["price_dollar"].value_counts().sort_index(ascending=False).head(10)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
Name: price_dollar, dtype: int64

In [14]:
autos["price_dollar"].value_counts().sort_index().head(10)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
Name: price_dollar, dtype: int64

**Outliers are outright visible. Let's use IQR score to find the outliers and turn them to null.**

In [15]:
Q1 = 1100
Q3 = 7200
IQR = 7200-1100

In [16]:
outliers = autos[(autos["price_dollar"] < (Q1 - 1.5 * IQR)) | (autos["price_dollar"] > (Q3 + 1.5 * IQR))]

In [17]:
outliers.shape

(3784, 20)

In [18]:
autos.loc[(autos["price_dollar"] < (Q1 - 1.5 * IQR)) | (autos["price_dollar"] > (Q3 + 1.5 * IQR)),"price_dollar"] = np.nan

** Lets also replace the $0 prices with null, since there are only 1421 cars which is only
2% of the of the cars. **

In [19]:
autos.loc[autos["price_dollar"] == 0 ,"price_dollar"] = np.nan

In [20]:
autos["price_dollar"].value_counts().sort_index(ascending=False).head(10)

16350.0     9
16333.0     2
16300.0    11
16299.0     4
16290.0     3
16250.0     9
16200.0    17
16190.0     1
16150.0     2
16100.0     2
Name: price_dollar, dtype: int64

In [21]:
autos["price_dollar"].value_counts().sort_index().head(10)

1.0     156
2.0       3
3.0       1
5.0       2
8.0       1
9.0       1
10.0      7
11.0      2
12.0      3
13.0      2
Name: price_dollar, dtype: int64

In [22]:
autos["price_dollar"].describe()

count    44795.000000
mean      4089.433620
std       3841.429247
min          1.000000
25%       1150.000000
50%       2700.000000
75%       5999.000000
max      16350.000000
Name: price_dollar, dtype: float64

** Using IQR method for turning outliers in 'odometer_km' to null. **

In [23]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [24]:
outliers = autos[(autos["odometer_km"] < (125000 - 1.5 * 25000)) | (autos["odometer_km"] > (150000 + 1.5 * 25000))]

In [25]:
outliers.shape

(8480, 20)

In [26]:
autos.loc[(autos["odometer_km"] < (125000 - 1.5 * 25000)) | (autos["odometer_km"] > (150000 + 1.5 * 25000)),"odometer_km"] = np.nan

In [27]:
autos = autos[~((autos["odometer_km"] < (125000 - 1.5*25000)) | (autos["odometer_km"] > (150000 + 1.5*25000)))]

In [28]:
autos.shape

(50000, 20)

**Outliers in prices and odometer were atleast ~20% of the data and hence the better option was to turn the outliers to null than to remove them. By further analysis of other columns we can find the rows with most null values i.e. most outliers and then remove them, thereby removing minimum amount of data.**

In [29]:
date_crawled = autos['date_crawled'].str[:4].astype(int)
print(date_crawled.describe())
print(autos['date_crawled'].isnull().value_counts())

count    50000.0
mean      2016.0
std          0.0
min       2016.0
25%       2016.0
50%       2016.0
75%       2016.0
max       2016.0
Name: date_crawled, dtype: float64
False    50000
Name: date_crawled, dtype: int64


In [30]:
date_created_year = autos["date_created"].str[:4].astype(int)
date_created_month = autos["date_created"].str[5:7].astype(int)
print(autos["date_created"].describe())
print(autos['date_created'].isnull().value_counts())

count                   50000
unique                     76
top       2016-04-03 00:00:00
freq                     1946
Name: date_created, dtype: object
False    50000
Name: date_created, dtype: int64


In [31]:
last_seen = autos['date_crawled'].str[:4].astype(int)
print(last_seen.describe())
print(autos['last_seen'].isnull().value_counts())

count    50000.0
mean      2016.0
std          0.0
min       2016.0
25%       2016.0
50%       2016.0
75%       2016.0
max       2016.0
Name: date_crawled, dtype: float64
False    50000
Name: last_seen, dtype: int64


In [32]:
print(autos['year_of_registration'].describe())
print(autos['year_of_registration'].isnull().value_counts())
autos['year_of_registration'].sort_values()
autos['year_of_registration'] = autos['year_of_registration'].astype(int)

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64
False    50000
Name: year_of_registration, dtype: int64


In [33]:
print(autos['month_of_registration'].describe())
print(autos['month_of_registration'].isnull().value_counts())
autos['month_of_registration'] = autos['month_of_registration'].astype(int)

count    50000.000000
mean         5.723360
std          3.711984
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: month_of_registration, dtype: float64
False    50000
Name: month_of_registration, dtype: int64


### ** Exploring Dates **

**The columns: ** *'date_crawled', 'last_seen', and 'date_created'* ** have no null values and have no outliers. **

** On the other hand the ** * 'month_of_registration' , 'year_of_registration'* ** have outliers which is evidendent from the descriptive statistics of the columns. **


** Making the dates of registration null that fall after the date the advertisement was created **

In [34]:
autos.loc[~((autos["year_of_registration"] < date_created_year) & (autos["month_of_registration"] < date_created_month)),"year_of_registration"] = np.nan
autos.loc[~((autos["year_of_registration"] < date_created_year) & (autos["month_of_registration"] < date_created_month)),"month_of_registration"] = np.nan

In [35]:
print(autos['year_of_registration'].describe())
print(autos['year_of_registration'].isnull().value_counts())

count    10901.000000
mean      2000.766077
std         18.737656
min       1000.000000
25%       1998.000000
50%       2001.000000
75%       2006.000000
max       2015.000000
Name: year_of_registration, dtype: float64
True     39099
False    10901
Name: year_of_registration, dtype: int64


** Removing outliers from month and year of registration columns: **

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate.Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

In [36]:
outliers = autos[(autos["year_of_registration"] < 1900)]

In [37]:
outliers.shape

(6, 20)

In [38]:
autos.loc[(autos["year_of_registration"] < 1900),"year_of_registration"] = np.nan

In [39]:
autos.shape

(50000, 20)

Descriptive stats of the month of registration suggests that the range of the column falls between 0-12. Thus, months less than 1 are outliers.

In [40]:
outliers = autos[(autos["month_of_registration"] < 1)]

In [41]:
outliers.shape

(4249, 20)

In [42]:
autos.loc[(autos["month_of_registration"] < 1),"month_of_registration"] = np.nan

In [43]:
autos.shape

(50000, 20)

** Since the outliers are easily identifiable and can be removed using a single boolean statement, using IQR wasn't necessary. **

** Finding outliesrs in 'power_ps' and 'postal_code':  **

In [44]:
autos["power_ps"].describe()

count    50000.000000
mean       116.355920
std        209.216627
min          0.000000
25%         70.000000
50%        105.000000
75%        150.000000
max      17700.000000
Name: power_ps, dtype: float64

In [45]:
outliers = autos[(autos["power_ps"] < (70 - 1.5 * 80)) | (autos["power_ps"] > (150 + 1.5 * 80))]

In [46]:
outliers.shape

(1461, 20)

In [47]:
autos.loc[(autos["power_ps"] < (70 - 1.5 * 80)) | (autos["power_ps"] > (150 + 1.5 * 80)),"power_ps"] = np.nan

In [48]:
autos = autos[~((autos["power_ps"] < (70 - 1.5*80)) | (autos["power_ps"] > (150 + 1.5*80)))]

In [49]:
autos.shape

(50000, 20)

In [50]:
autos["postal_code"].describe()

count    50000.000000
mean     50813.627300
std      25779.747957
min       1067.000000
25%      30451.000000
50%      49577.000000
75%      71540.000000
max      99998.000000
Name: postal_code, dtype: float64

**The postal codes fall in the range of 4-5 digits and thus, seem definite.**

In [51]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled             50000 non-null object
name                     50000 non-null object
seller                   50000 non-null object
offer_type               50000 non-null object
price_dollar             44795 non-null float64
abtest                   50000 non-null object
vehicle_type             44905 non-null object
year_of_registration     10895 non-null float64
gearbox                  47320 non-null object
power_ps                 48539 non-null float64
model                    47242 non-null object
odometer_km              41520 non-null float64
month_of_registration    6652 non-null float64
fuel_type                45518 non-null object
brand                    50000 non-null object
not_repaired_damage      40171 non-null object
date_created             50000 non-null object
nr_of_pictures           50000 non-null int64
postal_code              50000 non-null in

** There are 11 columns which include Null values. It is now appropriate to remove the rows which have 25% or more null values.**

In [52]:
autos = autos[autos.isnull().sum(axis=1) < 5]
autos.shape

(46877, 20)

### Analyzing and Exploring the Top Car Brands

** Top 10 common brands ** : Let's look at the 10 most common brands and the mean prices of these brands.

In [53]:
autos['brand'].value_counts().head(10)

volkswagen       10028
bmw               5138
opel              5132
mercedes_benz     4504
audi              4069
ford              3243
renault           2279
peugeot           1386
fiat              1213
seat               878
Name: brand, dtype: int64

In [66]:
top_brands = autos['brand'].value_counts().head(10).index

In [55]:
mean_price_brands = {}
for t in top_brands:
    if t not in mean_price_brands:
        mean_price_brands[t] = autos.loc[(autos["brand"] == t),"price_dollar"].mean()      

In [56]:
mean_price_brands

{'audi': 5734.181872391175,
 'bmw': 5678.881447963801,
 'fiat': 2770.9713080168776,
 'ford': 3001.188953488372,
 'mercedes_benz': 5282.397574479304,
 'opel': 2735.382347000606,
 'peugeot': 3005.9549815498153,
 'renault': 2288.97229791099,
 'seat': 3796.467703349282,
 'volkswagen': 4235.330335861322}

In [57]:
autos["price_dollar"].mean()

4164.371346396671

**Observations**
* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between
* Audi and BMW have the highest mean price of about \$5500 and Opel having the lowest mean, approx. \$2200. With Opel and BMW being more common than audi.
* Volkswagen, having the most postings, has a mean value little lower than the overall mean of all brands, ~$44114



** Most common brand models **

In [58]:
autos['brand_model']=autos['brand']+'_'+autos['model']

In [59]:
autos['brand_model'].value_counts().head(10)

volkswagen_golf           3824
bmw_3er                   2684
volkswagen_polo           1664
opel_corsa                1640
opel_astra                1392
volkswagen_passat         1388
audi_a4                   1265
mercedes_benz_c_klasse    1146
bmw_5er                   1143
mercedes_benz_e_klasse     963
Name: brand_model, dtype: int64

In [67]:
brand_model = autos['brand_model'].value_counts().head(10).index
mean_price_models = {}
for t in brand_model:
    if t not in mean_price_models:
        mean_price_models[t] = autos.loc[(autos["brand_model"] == t),"price_dollar"].mean()  

In [61]:
mean_price_models

{'audi_a4': 5612.73526824978,
 'bmw_3er': 4696.6727791361045,
 'bmw_5er': 5858.717479674797,
 'mercedes_benz_c_klasse': 4991.639134709932,
 'mercedes_benz_e_klasse': 5111.538271604938,
 'opel_astra': 3202.825958702065,
 'opel_corsa': 1861.480625,
 'volkswagen_golf': 4031.5540807681446,
 'volkswagen_passat': 4350.888632872504,
 'volkswagen_polo': 2547.4383477188658}

**Observations**
* Audi, BMW, Mercedes Benz, Opel and Volkswagen have the most common models
* Mercedes Benz has a high price but its models are the least common compared to the rest 
* Opel's models lie in the middle of the list being the least expensive and the most common after BMW and Volkswagen. Opel might be a 'best of both worlds' option.

Let's see if there's a visible link between the average mileage and the mean price of the top brands and models.

In [62]:
avg_mileage = {}
for t in brand_model:
    if t not in avg_mileage:
        avg_mileage[t] = autos.loc[(autos["brand_model"] == t),"odometer_km"].mean()  

In [63]:
avg_mileage

{'audi_a4': 144913.64421416234,
 'bmw_3er': 144806.6958947788,
 'bmw_5er': 146014.6252285192,
 'mercedes_benz_c_klasse': 143257.804632427,
 'mercedes_benz_e_klasse': 145826.72706681766,
 'opel_astra': 142361.78861788617,
 'opel_corsa': 140266.80526680526,
 'volkswagen_golf': 143013.42691486116,
 'volkswagen_passat': 146247.13083397094,
 'volkswagen_polo': 141466.06018194542}

* Comparing avg mileage and mean prices of the top models:

In [79]:
prices = pd.Series(mean_price_models).sort_values(ascending=False) 
mileage = pd.Series(avg_mileage).sort_values(ascending=False)
info = pd.DataFrame(mileage,columns=['mean_mileage'])

In [80]:
info["mean_price"] = prices
info

mean_mileage   mean_price
volkswagen_passat       146247.130834  4350.888633
bmw_5er                 146014.625229  5858.717480
mercedes_benz_e_klasse  145826.727067  5111.538272
audi_a4                 144913.644214  5612.735268
bmw_3er                 144806.695895  4696.672779
mercedes_benz_c_klasse  143257.804632  4991.639135
volkswagen_golf         143013.426915  4031.554081
opel_astra              142361.788618  3202.825959
volkswagen_polo         141466.060182  2547.438348
opel_corsa              140266.805267  1861.480625

There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage. But overall, the range of car mileages does not vary as much as the prices do by brand.